In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
categorical_path = 'datasets/chronic_kidney_disease_categorical.csv'
numerical_path = 'datasets/chronic_kidney_disease_numerical.csv'

categorical_data = pd.read_csv(categorical_path)
numerical_data = pd.read_csv(numerical_path)

In [3]:
# Print the data
categorical_data

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


In [4]:
numerical_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


In [5]:
print(f'categorical data keys: {list(categorical_data.keys())}')

categorical data keys: ['unique_id', 'al', 'su', 'rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']


### Categorical data keys meaning

| variable name | description | type |
| :-----------: | :---------: | :--: |
| al | albumin | categorical |
| su | sugar | categorical |
| rbc | red blood cells | binary |
| pc | pus cell | binary |
| pcc | pus cell pumps | binary |
| ba | bacteria | binary |
| htn | hypertension | binary |
| dm | diabetes mellitus | binary |
| cad | coronary artery disease | binary |
| appet | appetite | binary |
| pe | pedal edema | binary |
| ane | anemia | binary |
| target | class, ckd or not | binary |


In [6]:
print(f'numerical data keys: {list(numerical_data.keys())}')

numerical data keys: ['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']


### Numerical data keys meaning

| variable name | description | type |
| :-----------: | :---------: | :--: |
| age | age | int |
| bp | blood pressure| int |
| bgr | blood glucose random | int|
| bu | blood urea | int |
| sc | serum creatinine | cont |
| sod | sodium | int |
| pot | potassium | cont | 
| hemo | hemoglobin | cont |
| pcv | packed cell volume | cont |
| wbcc | white blood cell count | int |
| rbcc | red blood cell count | count |

In [7]:
print(f'numerical_data length: {len(numerical_data)}')
print(f'categorical_data length: {len(categorical_data)}')

numerical_data length: 420
categorical_data length: 430


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

**Answer:**

*This dataset is separated into two data frames, one containing all categorical features and the other containing all numerical features. Each feature is a factor that might indicate CKD to some extent. Each patient is assigned to a unique id that matches between the numerical df and categorical df (although by the looks of it not all of them match). Some examples of the features from the categorical data frame include albumin, sugar, and pus cell. Some examples from the numerical data frame include age, blood pressure, and blood glucose random.*

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [8]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print(f'Number of duplicates, Numerical dataset: {numerical_data['unique_id'].duplicated().sum()}')

# Drop these duplicate rows
numerical_data = numerical_data.drop_duplicates('unique_id')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(f'Number of duplicates, Categorical dataset: {categorical_data['unique_id'].duplicated().sum()}')
categorical_data = categorical_data.drop_duplicates('unique_id')

Number of duplicates, Numerical dataset: 30
Number of duplicates, Categorical dataset: 30


In [9]:
len(numerical_data) == len(categorical_data)

False

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: *No. The length of numerical data is not equal to categorical data, so it is impossible that all unique ids are present in both. When combining, these rows will have missing categorical or numerical data.*

In [10]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = pd.merge(numerical_data, categorical_data, on='unique_id')

#Print the combined dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [11]:
# Calculate the percentage of rows that contain at least one missing value
percentage_nan = float(merged.isna().any(axis=1).sum() / len(merged))
# Print %
print(f'Percentage of rows with missing value: {percentage_nan:.4f}')

# Drop these rows from the dataset
merged = merged.dropna().reset_index(drop=True)
# Print the Dataset
merged

Percentage of rows with missing value: 0.6077


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
4,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
151,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [12]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged = merged.sort_values(by=['Target']).reset_index(drop=True)

# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this**

In [13]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
merged = pd.get_dummies(merged, categorical_columns, dtype=int)
# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,1,0,1,0,1,0,0,1,1,0
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,0,1,0,1,1,0,1,0,1,0
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,1,0,0,1,1,0,0,1,1,0
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,1,0,1,0,1,0,1,0,0,1
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,1,0,1,0,1,0,1,0,0,1
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,1,0,1,0,1,0,1,0,0,1
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [14]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
Q1 = merged[numerical_columns].quantile(0.9973)
Q2 = merged[numerical_columns].quantile(1 - 0.9973)

merged = merged[((merged[numerical_columns] >= Q2) & (merged[numerical_columns] <= Q1)).all(axis=1)].reset_index(drop=True)
# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,1,0,1,0,1,0,0,1,1,0
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,0,1,0,1,1,0,1,0,1,0
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,1,0,0,1,1,0,0,1,1,0
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,1,0,1,0,1,0,1,0,0,1
137,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,1,0,1,0,1,0,1,0,0,1
138,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,1,0,1,0,1,0,1,0,0,1
139,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [15]:
# Normalize the all Numerical Attributes in the dataset.
merged[numerical_columns] = (merged[numerical_columns] - merged[numerical_columns].min()) / (merged[numerical_columns].max() - merged[numerical_columns].min())
# Print the dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,0.132353,0.75,0.104520,0.133333,0.087591,0.305556,0.171429,0.221311,0.184211,...,1,0,1,0,1,0,0,1,1,0
1,137148,0.897059,1.00,0.635593,0.355556,0.372263,0.722222,0.000000,0.295082,0.368421,...,0,1,0,1,1,0,1,0,1,0
2,484175,0.602941,0.75,0.000000,0.431111,0.489051,0.000000,0.228571,0.319672,0.342105,...,1,0,0,1,1,0,0,1,1,0
3,343710,0.720588,0.50,0.290960,0.613333,0.248175,0.583333,0.657143,0.172131,0.210526,...,0,1,0,1,0,1,0,1,1,0
4,532520,0.705882,0.75,0.098870,0.191111,0.131387,0.611111,0.657143,0.450820,0.447368,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136,647552,0.720588,0.25,0.141243,0.084444,0.014599,0.638889,0.171429,0.967213,0.947368,...,1,0,1,0,1,0,1,0,0,1
137,125587,0.367647,0.00,0.110169,0.164444,0.043796,0.750000,0.571429,0.770492,0.842105,...,1,0,1,0,1,0,1,0,0,1
138,614376,0.661765,0.50,0.138418,0.031111,0.051095,0.583333,0.514286,0.803279,0.684211,...,1,0,1,0,1,0,1,0,0,1
139,853426,0.426471,0.25,0.155367,0.124444,0.007299,0.722222,0.600000,0.918033,0.657895,...,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: *The unique_id is no longer useful as it only served aa an identifier before the data frame was merged. Target_notckd is also excessive as Target_ckd itself can indicate whether ckd is presented.*

In [16]:
#Remove that column
merged = merged.drop(['unique_id', 'Target_notckd'], axis=1)
# Print the dataset
merged

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd
0,0.132353,0.75,0.104520,0.133333,0.087591,0.305556,0.171429,0.221311,0.184211,0.547297,...,0,1,0,1,0,1,0,0,1,1
1,0.897059,1.00,0.635593,0.355556,0.372263,0.722222,0.000000,0.295082,0.368421,0.182432,...,1,0,1,0,1,1,0,1,0,1
2,0.602941,0.75,0.000000,0.431111,0.489051,0.000000,0.228571,0.319672,0.342105,0.527027,...,1,1,0,0,1,1,0,0,1,1
3,0.720588,0.50,0.290960,0.613333,0.248175,0.583333,0.657143,0.172131,0.210526,0.331081,...,1,0,1,0,1,0,1,0,1,1
4,0.705882,0.75,0.098870,0.191111,0.131387,0.611111,0.657143,0.450820,0.447368,0.418919,...,0,1,0,1,0,1,0,1,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136,0.720588,0.25,0.141243,0.084444,0.014599,0.638889,0.171429,0.967213,0.947368,0.182432,...,0,1,0,1,0,1,0,1,0,0
137,0.367647,0.00,0.110169,0.164444,0.043796,0.750000,0.571429,0.770492,0.842105,0.182432,...,0,1,0,1,0,1,0,1,0,0
138,0.661765,0.50,0.138418,0.031111,0.051095,0.583333,0.514286,0.803279,0.684211,0.128378,...,0,1,0,1,0,1,0,1,0,0
139,0.426471,0.25,0.155367,0.124444,0.007299,0.722222,0.600000,0.918033,0.657895,0.135135,...,0,1,0,1,0,1,0,1,0,0


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [17]:
# Export the dataframe to a new csv file
import os.path as osp
dir = './datasets_clean'
merged.to_csv(osp.join(dir,'chronic_kidney_disease.csv'))
# Export the dataframe to a new json file
merged.to_json(osp.join(dir,'chronic_kidney_disease.json'))

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [18]:
pd.set_option('display.max_columns', None)
merged.head(15)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd
0,0.132353,0.75,0.104520,0.133333,0.087591,0.305556,0.171429,0.221311,0.184211,0.547297,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1
1,0.897059,1.00,0.635593,0.355556,0.372263,0.722222,0.000000,0.295082,0.368421,0.182432,0.153846,3.0,2.0,1,0,1,0,0,1,1,0,0,1,0,1,0,1,0,1,1,0,1,0,1
2,0.602941,0.75,0.000000,0.431111,0.489051,0.000000,0.228571,0.319672,0.342105,0.527027,0.282051,2.0,0.0,1,0,1,0,0,1,1,0,0,1,0,1,1,0,0,1,1,0,0,1,1
3,0.720588,0.50,0.290960,0.613333,0.248175,0.583333,0.657143,0.172131,0.210526,0.331081,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1
4,0.705882,0.75,0.098870,0.191111,0.131387,0.611111,0.657143,0.450820,0.447368,0.418919,0.384615,2.0,0.0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1
5,0.544118,1.00,0.166667,0.657778,0.824818,0.222222,0.085714,0.204918,0.210526,0.358108,0.230769,3.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,1,0,0,1,0,1,0,1,1
6,0.500000,0.00,0.262712,0.364444,0.204380,0.750000,0.314286,0.344262,0.315789,0.695946,0.153846,1.0,0.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1
7,0.632353,0.75,0.573446,1.000000,1.000000,0.500000,0.142857,0.221311,0.157895,0.695946,0.076923,2.0,1.0,1,0,1,0,1,0,1,0,0,1,0,1,1,0,0,1,0,1,0,1,1
8,0.647059,0.75,0.166667,0.431111,0.452555,0.472222,0.542857,0.286885,0.342105,0.141892,0.205128,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1
9,0.691176,0.50,0.658192,0.111111,0.058394,0.222222,0.171429,0.393443,0.500000,0.445946,0.435897,1.0,0.0,1,0,0,1,1,0,1,0,1,0,0,1,1,0,0,1,1,0,1,0,1


|    | age       | bp   | bgr      | bu       | sc       | sod      | pot      | hemo     | pcv      | wbcc     | rbcc     | al    | su    | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|-----------|------|----------|----------|----------|----------|----------|----------|----------|----------|----------|-------|-------|---------------|------------|-------------|-----------|-----------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|-------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0  | 0.132353  | 0.75 | 0.104520 | 0.133333 | 0.087591 | 0.305556 | 0.171429 | 0.221311 | 0.184211 | 0.547297 | 0.333333 | 4.0   | 0.0   | 0             | 1          | 1           | 0         | 0               | 1           | 0              | 1          | 1      | 0       | 1     | 0      | 1      | 1       | 0           | 1          | 0     | 1      | 0      | 1       | 0           | 0              |
| 1  | 0.897059  | 1.00 | 0.635593 | 0.355556 | 0.372263 | 0.722222 | 0.000000 | 0.295082 | 0.368421 | 0.182432 | 0.153846 | 3.0   | 2.0   | 1             | 0          | 1           | 0         | 0               | 1           | 1              | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 0     | 1      | 1      | 0       | 1           | 0              |
| 2  | 0.602941  | 0.75 | 0.000000 | 0.431111 | 0.489051 | 0.000000 | 0.228571 | 0.319672 | 0.342105 | 0.527027 | 0.282051 | 2.0   | 0.0   | 1             | 0          | 1           | 0         | 0               | 1           | 1              | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 0           | 1          | 1     | 0      | 0      | 1       | 1           | 0              |
| 3  | 0.720588  | 0.50 | 0.290960 | 0.613333 | 0.248175 | 0.583333 | 0.657143 | 0.172131 | 0.210526 | 0.331081 | 0.153846 | 2.0   | 0.0   | 1             | 0          | 1           | 0         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 0           | 1          | 0     | 1      | 0      | 1       | 0           | 1              |
| 4  | 0.705882  | 0.75 | 0.098870 | 0.191111 | 0.131387 | 0.611111 | 0.657143 | 0.450820 | 0.447368 | 0.418919 | 0.384615 | 2.0   | 0.0   | 1             | 0          | 0           | 1         | 1               | 0           | 1              | 0          | 1      | 0       | 1     | 0      | 1      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 5  | 0.544118  | 1.00 | 0.166667 | 0.657778 | 0.824818 | 0.222222 | 0.085714 | 0.204918 | 0.210526 | 0.358108 | 0.230769 | 3.0   | 0.0   | 1             | 0          | 1           | 0         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 1           | 0          | 0     | 1      | 0      | 1       | 1           | 0              |
| 6  | 0.500000  | 0.00 | 0.262712 | 0.364444 | 0.204380 | 0.750000 | 0.314286 | 0.344262 | 0.315789 | 0.695946 | 0.153846 | 1.0   | 0.0   | 0             | 1          | 0           | 1         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 7  | 0.632353  | 0.75 | 0.573446 | 1.000000 | 1.000000 | 0.500000 | 0.142857 | 0.221311 | 0.157895 | 0.695946 | 0.076923 | 2.0   | 1.0   | 1             | 0          | 1           | 0         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 1           | 0          | 0     | 1      | 0      | 1       | 1           | 0              |
| 8  | 0.647059  | 0.75 | 0.166667 | 0.431111 | 0.452555 | 0.472222 | 0.542857 | 0.286885 | 0.342105 | 0.141892 | 0.205128 | 2.0   | 0.0   | 1             | 0          | 1           | 0         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 1      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 9  | 0.691176  | 0.50 | 0.658192 | 0.111111 | 0.058394 | 0.222222 | 0.171429 | 0.393443 | 0.500000 | 0.445946 | 0.435897 | 1.0   | 0.0   | 1             | 0          | 0           | 1         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 1      | 0       | 0           | 1          | 1     | 0      | 0      | 1       | 1           | 0              |
| 10 | 0.897059  | 0.25 | 0.000000 | 0.097778 | 0.029197 | 0.305556 | 0.314286 | 0.360656 | 0.342105 | 0.986486 | 0.230769 | 0.0   | 0.0   | 0             | 1          | 0           | 1         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 1           | 0          | 1     | 0      | 0      | 1       | 1           | 0              |
| 11 | 0.705882  | 0.00 | 0.615819 | 0.115556 | 0.087591 | 0.444444 | 0.028571 | 0.188525 | 0.236842 | 0.736486 | 0.102564 | 3.0   | 1.0   | 0             | 1          | 1           | 0         | 0               | 1           | 1              | 0          | 0      | 1       | 1     | 0      | 1      | 1       | 0           | 1          | 0     | 0      | 1      | 1       | 0           | 0              |
| 12 | 0.823529  | 0.50 | 0.245763 | 0.355556 | 0.262774 | 0.444444 | 1.000000 | 0.000000 | 0.000000 | 0.452703 | 0.000000 | 3.0   | 2.0   | 0             | 1          | 1           | 0         | 0               | 1           | 0              | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 0     | 1      | 0      | 1       | 0           | 1              |
| 13 | 0.867647  | 0.00 | 0.135593 | 0.511111 | 0.350365 | 0.611111 | 0.571429 | 0.475410 | 0.500000 | 0.736486 | 0.435897 | 4.0   | 0.0   | 0             | 1          | 0           | 1         | 1               | 0           | 1              | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 1           | 0          | 1     | 0      | 0      | 1       | 1           | 0              |
| 14 | 0.838235  | 0.25 | 0.406780 | 0.382222 | 0.423358 | 0.166667 | 0.285714 | 0.311475 | 0.315789 | 0.486486 | 0.179487 | 4.0   | 3.0   | 0             | 1          | 1           | 0         | 0               | 1           | 0              | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 0     | 1      | 1      | 0       | 0           | 1              |


**Caution:** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

- implement checks to verify the conversion.
- break-down large data into chunks
- test with known data to check performance
- logging and monitoring while data conversion happens
- fine-tune LLMs on specific datasets

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [19]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell pumps',
    'ba': 'bacteria',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'target': 'class, ckd or not',
    'age': 'age',
    'bp': 'blood pressure',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'Target_ckd': 'Target'
}
merged.rename(columns=rename_mapping, inplace=True)
merged

,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,white blood cell count,red blood cell count,albumin,sugar,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target
0,0.132353,0.75,0.104520,0.133333,0.087591,0.305556,0.171429,0.221311,0.184211,0.547297,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1
1,0.897059,1.00,0.635593,0.355556,0.372263,0.722222,0.000000,0.295082,0.368421,0.182432,0.153846,3.0,2.0,1,0,1,0,0,1,1,0,0,1,0,1,0,1,0,1,1,0,1,0,1
2,0.602941,0.75,0.000000,0.431111,0.489051,0.000000,0.228571,0.319672,0.342105,0.527027,0.282051,2.0,0.0,1,0,1,0,0,1,1,0,0,1,0,1,1,0,0,1,1,0,0,1,1
3,0.720588,0.50,0.290960,0.613333,0.248175,0.583333,0.657143,0.172131,0.210526,0.331081,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1
4,0.705882,0.75,0.098870,0.191111,0.131387,0.611111,0.657143,0.450820,0.447368,0.418919,0.384615,2.0,0.0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136,0.720588,0.25,0.141243,0.084444,0.014599,0.638889,0.171429,0.967213,0.947368,0.182432,0.692308,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0
137,0.367647,0.00,0.110169,0.164444,0.043796,0.750000,0.571429,0.770492,0.842105,0.182432,0.666667,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0
138,0.661765,0.50,0.138418,0.031111,0.051095,0.583333,0.514286,0.803279,0.684211,0.128378,0.923077,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0
139,0.426471,0.25,0.155367,0.124444,0.007299,0.722222,0.600000,0.918033,0.657895,0.135135,0.846154,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [20]:
# Converted SQL to Pandas code
result = merged.groupby('Target').size().reset_index(name='count')
result

,Target,count
0,0,110
1,1,31


*It groups every unique value in the Target column, calculates the size of each group to get the frequency of each value, then displays the frequency in the column count*